Source/Alamofire.swift

```diff
@@ -27,42 +27,83 @@ import Foundation
 /// Types adopting the `URLStringConvertible` protocol can be used to construct URL strings, which are then used to
 /// construct URL requests.
 public protocol URLStringConvertible {
-    /// A URL that conforms to RFC 2396.
+    /// A URL string that conforms to RFC 2396.
     ///
     /// Methods accepting a `URLStringConvertible` type parameter parse it according to RFCs 1738 and 1808.
     ///
     /// See https://tools.ietf.org/html/rfc2396
     /// See https://tools.ietf.org/html/rfc1738
     /// See https://tools.ietf.org/html/rfc1808
-    var urlString: String { get }
+    var urlString: String? { get }
+
+    /// Returns a URL string that conforms to RFC 2396 or throws if an `Error` was encoutered.
+    ///
+    /// Methods accepting a `URLStringConvertible` type parameter parse it according to RFCs 1738 and 1808.
+    ///
+    /// - throws: An `Error` if underlying url string is `nil`.
+    ///
+    /// - returns: A URL string.
+    func asURLString() throws -> String
+}
+
+extension URLStringConvertible {
+    public var urlString: String? { return try? asURLString() }
 }
 
 extension String: URLStringConvertible {
-    /// The URL string.
-    public var urlString: String { return self }
+    /// Returns a URL string that conforms to RFC 2396.
+    public func asURLString() throws -> String { return self }
 }
 
 extension URL: URLStringConvertible {
-    /// The URL string.
-    public var urlString: String { return absoluteString }
+    /// Returns a URL string that conforms to RFC 2396.
+    public func asURLString() throws -> String { return absoluteString }
+
+    /// Creates a URL with the specified URL string if possible, otherwise throws an `Error`.
+    ///
+    /// - parameter urlString: The URL string convertible to create the URL with.
+    ///
+    /// - throws: An `AFError.invalidURLString` if invalid.
+    ///
+    /// - returns: A URL if created successfully, otherwise throws an `AFError`.
+    public init(urlString: URLStringConvertible) throws {
+        let urlString = try urlString.asURLString()
+        guard let url = URL(string: urlString) else { throw AFError.invalidURLString(urlString: urlString) }
+        self = url
+    }
 }
 
 extension URLComponents: URLStringConvertible {
-    /// The URL string.
-    public var urlString: String { return url!.urlString }
+    /// Returns a URL string that conforms to RFC 2396.
+    public func asURLString() throws -> String {
+        guard let urlString = url?.urlString else { throw AFError.invalidURLString(urlString: self) }
+        return urlString
+    }
 }
 
 // MARK: -
 
 /// Types adopting the `URLRequestConvertible` protocol can be used to construct URL requests.
 public protocol URLRequestConvertible {
     /// The URL request.
-    var urlRequest: URLRequest { get }
+    var urlRequest: URLRequest? { get }
```

Comment:
Same as with `URLStringConvertible`, I don't think we need the property in the protocol anymore.